middleware/commandhandler/scheduler/middleware.go

```diff
@@ -82,6 +52,302 @@ func (c *command) ExecuteAt() time.Time {
+				_, err := s.ScheduleCommand(ctx, cmd, c.ExecuteAt())
+
+				return err
+			}
+
+			// Immediate command execution.
+			return h.HandleCommand(ctx, cmd)
+		})
+	}), s
+}
+
+// PersistedCommand is a persisted command.
+type PersistedCommand struct {
+	ID         uuid.UUID       `json:"_"                 bson:"_id"`
+	IDStr      string          `json:"id"                bson:"_"`
+	RawCommand []byte          `json:"command"           bson:"command"`
+	ExecuteAt  time.Time       `json:"timestamp"         bson:"timestamp"`
+	Command    eh.Command      `json:"-"                 bson:"-"`
+	Context    context.Context `json:"-"                 bson:"-"`
+}
+
+// EntityID implements the EntityID method of the eventhorizon.Entity interface.
+func (c *PersistedCommand) EntityID() uuid.UUID {
+	return c.ID
+}
+
+// Scheduler is a scheduled of commands.
+type Scheduler struct {
+	h                  eh.CommandHandler
+	hMu                sync.Mutex
+	repo               eh.ReadWriteRepo
+	cmdCh              chan *scheduledCommand
+	cancelScheduling   map[uuid.UUID]chan struct{}
+	cancelSchedulingMu sync.Mutex
+	errCh              chan error
+	cctx               context.Context
+	cancel             context.CancelFunc
+	done               chan struct{}
+	codec              eh.CommandCodec
+}
+
+func (s *Scheduler) setHandler(h eh.CommandHandler) {
+	s.hMu.Lock()
+	defer s.hMu.Unlock()
+
+	s.h = h
+}
+
+// Start starts the scheduler by first loading all persisted commands.
+func (s *Scheduler) Start() error {
+	if err := s.loadCommands(); err != nil {
+		return fmt.Errorf("could not load commands: %w", err)
+	}
+
+	s.cctx, s.cancel = context.WithCancel(context.Background())
+	s.done = make(chan struct{})
+
+	go s.run()
+
+	return nil
+}
+
+// Stop stops all scheduled commands.
+func (s *Scheduler) Stop() error {
+	s.cancel()
+
+	<-s.done
+
+	return nil
+}
+
+// Errors returns an error channel that will receive errors from handling of
+// scheduled commands.
+func (s *Scheduler) Errors() <-chan error {
+	return s.errCh
+}
+
+type scheduledCommand struct {
+	id        uuid.UUID
+	ctx       context.Context
+	cmd       eh.Command
+	executeAt time.Time
+}
+
+// ScheduleCommand schedules a command to be executed at `executeAt`. It is persisted
+// to the repo.
+func (s *Scheduler) ScheduleCommand(ctx context.Context, cmd eh.Command, executeAt time.Time) (uuid.UUID, error) {
+	b, err := s.codec.MarshalCommand(ctx, cmd)
+	if err != nil {
+		return uuid.Nil, &Error{
+			Err:     fmt.Errorf("could not marshal command: %w", err),
+			Ctx:     ctx,
+			Command: cmd,
+		}
+	}
+
+	// Use the command ID as persisted ID if available.
+	var id uuid.UUID
+	if cmd, ok := cmd.(eh.CommandIDer); ok {
+		id = cmd.CommandID()
+	} else {
+		id = uuid.New()
+	}
+
+	pc := &PersistedCommand{
+		ID:         id,
+		IDStr:      id.String(),
+		RawCommand: b,
+		ExecuteAt:  executeAt,
+	}
+
+	if err := s.repo.Save(context.Background(), pc); err != nil {
```

Comment:
The context of the current namespace might be needed here?

Reply:
I'll fix this in this PR.